### ISSUE TITLE
Fast Garbage Collection

### ISSUE THREAD

**Issue by mist64** [commanderx16/x16-rom#155]
Microsoft BASIC for 6502 uses a notoriously slow (O(n^2)) garbage collection algorithm. Commodore replaced it with a faster version in later versions of BASIC, and there are several patches and ROM hacks for the C64 that also replace the algorithm.

We should pick one of these implementations and replace the existing algorithm with it.

**Comment by BruceMcF** [comment]
The benefit of this as a project from someone to take on is that the changes are atomic ... it involves the routine that places string data on the head putting an extra two bytes that link back to the variable that owns the data, and then modifying the garbage collector to take advantage of it.

Even without changing the variable allocation routine, the garbage collector itself could be sped up with access to more memory, as the reason it is so especially slow is it scans through the variables for each string rather than going through the string variable references and mapping the heap to find the missing spots. Building a BAM map of the heap would make garbage collection substantially faster, at the cost of space for the BAM in the Basic High RAM Segment ...
... my first guess is it would be linear like the C128 collector, though obviously a slower linear process.

**Comment by bobbyjim** [comment]
YES.  I was going to request this... once I learned that BASIC 4.0 fixed the problem, that is.

As Bruce said, this is an update on how strings are managed... but rather than a simple change, I suspect this could affect EVERYTHING in the kernal that touches strings.  Maybe.  That doesn't sound easy.  (On the other hand, if it is as easy as Bruce says, then HECK YES "we" should do this!!)

But the benefits are substantial.  Just to improve the GC is reason alone to consider this.   I mean, does the X16 **need** to lock up just to clean up orphan strings from the heap?  NO, I say!
